client/src/main/scala/org/apache/celeborn/client/commit/ReducePartitionCommitHandler.scala

```diff
@@ -0,0 +1,176 @@
+      logInfo(s"[handleStageEnd] Shuffle $shuffleId already ended!")
+      return
+    }
+    inProcessStageEndShuffleSet.synchronized {
+      if (inProcessStageEndShuffleSet.contains(shuffleId)) {
+        logWarning(s"[handleStageEnd] Shuffle $shuffleId is in process!")
+        return
+      }
+      inProcessStageEndShuffleSet.add(shuffleId)
+    }
+    // ask allLocations workers holding partitions to commit files
+    val shuffleAllocatedWorkers = allocatedWorkers.get(shuffleId)
+    val (dataLost, commitFailedWorkers) = handleFinalCommitFiles(shuffleId, shuffleAllocatedWorkers)
+    recordWorkerFailure(commitFailedWorkers)
+    // reply
+    if (!dataLost) {
+      logInfo(s"Succeed to handle stageEnd for $shuffleId.")
+      // record in stageEndShuffleSet
+      stageEndShuffleSet.add(shuffleId)
+    } else {
+      logError(s"Failed to handle stageEnd for $shuffleId, lost file!")
+      dataLostShuffleSet.add(shuffleId)
+      // record in stageEndShuffleSet
+      stageEndShuffleSet.add(shuffleId)
+    }
+    inProcessStageEndShuffleSet.remove(shuffleId)
+  }
+
+  private def handleFinalCommitFiles(
+      shuffleId: Int,
+      allocatedWorkers: util.Map[WorkerInfo, PartitionLocationInfo])
+      : (Boolean, ShuffleFailedWorkers) = {
+    val shuffleCommittedInfo = committedPartitionInfo.get(shuffleId)
+
+    // commit files
+    val parallelCommitResult = parallelCommitFiles(shuffleId, allocatedWorkers, None)
+
+    // check all inflight request complete
+    waitInflightRequestComplete(shuffleCommittedInfo)
+
+    // check data lost
+    val dataLost = checkDataLost(
+      shuffleId,
+      shuffleCommittedInfo.failedMasterPartitionIds,
+      shuffleCommittedInfo.failedSlavePartitionIds)
+
+    // collect result
+    if (!dataLost) {
+      collectResult(
+        shuffleId,
+        shuffleCommittedInfo,
+        getPartitionUniqueIds(shuffleCommittedInfo.committedMasterIds),
+        getPartitionUniqueIds(shuffleCommittedInfo.committedSlaveIds),
+        parallelCommitResult.masterPartitionLocationMap,
+        parallelCommitResult.slavePartitionLocationMap)
+    }
+
+    (dataLost, parallelCommitResult.commitFilesFailedWorkers)
+  }
+
+  override def finalPartitionCommit(
+      shuffleId: Int,
+      partitionId: Int,
+      recordWorkerFailure: ShuffleFailedWorkers => Unit): Boolean = {
+    throw new UnsupportedOperationException(
+      s"Failed when do final Partition Commit Operation, Reduce Partition " +
+        s"shuffleType only Support final commit for all partitions ")
+  }
+
+  override def getShuffleMapperAttempts(shuffleId: Int): Array[Int] = {
+    shuffleMapperAttempts.get(shuffleId)
+  }
+
+  def waitInflightRequestComplete(shuffleCommittedInfo: ShuffleCommittedInfo): Unit = {
+    while (shuffleCommittedInfo.allInFlightCommitRequestNum.get() > 0) {
+      Thread.sleep(1000)
+    }
+  }
+
+  def getPartitionUniqueIds(ids: ConcurrentHashMap[Int, util.List[String]])
```

Comment:
I don't think we should define this method

Reply:
It's just return an iterator to make code more clearer.